win32ss/user/user32/windows/input.c

```diff
@@ -657,11 +662,19 @@ IntLoadKeyboardLayout(
     UNICODE_STRING ustrKLID;
     WCHAR wszRegKey[256] = L"SYSTEM\\CurrentControlSet\\Control\\Keyboard Layouts\\";
     WCHAR wszLayoutId[10], wszNewKLID[10], szImeFileName[80];
+    PWCHAR endptr;
     HKL hNewKL;
     HKEY hKey;
     BOOL bIsIME;
 
-    dwhkl = wcstoul(pwszKLID, NULL, 16);
+    if (!IsValidKLID(pwszKLID))
```

Comment:
Have you tried, on Windows, whether specifying in the LoadKeyboardLayout (and having in the registry, a corresponding entry) a `pwszKLID` string whose format is not the one you enforce here, but instead a modified one:
`0xHEXNUMBER other_characters`
(here, `HEXNUMBER` is the 8-character long ID in hexadecimal),
where a hex prefix has been added and other characters follow the ID in hexadecimal (after a separating space) ?

Reply:
```txt
#include <windows.h>
#include <stdio.h>

int main(void)
{
    printf("%p\n", LoadKeyboardLayoutA("", KLF_REPLACELANG));
    printf("%p\n", LoadKeyboardLayoutA("<>", KLF_REPLACELANG));
    printf("%p\n", LoadKeyboardLayoutA("411   ", KLF_REPLACELANG));
    printf("%p\n", LoadKeyboardLayoutA("   411  ", KLF_REPLACELANG));
    printf("%p\n", LoadKeyboardLayoutA("000411", KLF_REPLACELANG));
    printf("%p\n", LoadKeyboardLayoutA("000411-", KLF_REPLACELANG));
    printf("%p\n", LoadKeyboardLayoutA("00000409", KLF_REPLACELANG));
    printf("%p\n", LoadKeyboardLayoutA("409", KLF_REPLACELANG));
    printf("%p\n", LoadKeyboardLayoutA("ffffffff", KLF_REPLACELANG));
}
```

![VirtualBox_Windows XP Japanese_04_09_2022_08_29_07](https://user-images.githubusercontent.com/2107452/188290564-cf7d869a-30f1-4aca-90c0-adb36ed1e7d9.png)
The function returns the meaningful value even if it failed (`04090409` is English).